bundles/org.openhab.binding.sony/src/main/java/org/openhab/binding/sony/internal/providers/sources/AbstractSonySource.java

```diff
@@ -0,0 +1,685 @@
+                                }));
+
+                        final List<ChannelGroupDefinition> gDefs = cgd.entrySet().stream()
+                                .map(gt -> new ChannelGroupDefinition(gt.getKey(), gt.getValue().getUID()))
+                                .collect(Collectors.toList());
+
+                        final URI configUri = new URI(configUriStr);
+                        final ThingType thingType = ThingTypeBuilder
+                                .instance(SonyBindingConstants.BINDING_ID, thingTypeId, label)
+                                .withConfigDescriptionURI(configUri).withDescription(desc)
+                                .withChannelGroupDefinitions(gDefs).build();
+
+                        final ThingTypeUID uid = thingType.getUID();
+
+                        groupTypes.putAll(cgd.values().stream().collect(Collectors.toMap(k -> k.getUID(), v -> v)));
+                        thingTypes.put(uid, thingType);
+                        thingTypeDefinitions.put(uid, ttd);
+
+                        results.add(new AbstractMap.SimpleEntry<>(thingType, ttd));
+
+                        fireThingTypeFound(uid);
+
+                        logger.debug("Successfully created a thing type {} from {}", thingType.getUID(), referenceName);
+
+                    } catch (final URISyntaxException e) {
+                        validationMessage.add("Configuration URI (" + configUriStr + ") was not a valid URI");
+                    }
+                }
+
+                if (!validationMessage.isEmpty()) {
+                    logger.debug("Error creating a thing type from element #{} ({}) in {}:", idx, modelName,
+                            referenceName);
+                    for (final String msg : validationMessage) {
+                        logger.debug("   {}", msg);
+                    }
+                }
+            }
+            return results;
+        } finally {
+            writeLock.unlock();
+        }
+    }
+
+    @Override
+    public void addListener(final String modelName, final ThingTypeUID currentThingTypeUID,
+            final SonyModelListener listener) {
+        Validate.notEmpty(modelName, "modelName cannot be empty");
+        Objects.requireNonNull(currentThingTypeUID, "currentThingTypeUID cannot be null");
+        Objects.requireNonNull(listener, "listener cannot be null");
+
+        final String serviceName = SonyUtil.getServiceName(currentThingTypeUID);
+        final ServiceModelName srvModelName = new ServiceModelName(serviceName, modelName);
+
+        final Lock writeLock = listenerLock.writeLock();
+        try {
+            writeLock.lock();
+
+            List<SonyModelListener> list = getListeners(srvModelName);
+            if (list == null) {
+                list = new ArrayList<>();
+                listeners.put(srvModelName, list);
+            }
+            if (!list.contains(listener)) {
+                list.add(listener);
+            }
```

Comment:
Because a hashset would need a listener to implement equals/hashcode and they don't - maybe I'm missing what you are saying...

Reply:
Assuming that your listener implementations don't implement `equals` (which I assume the service classes that implement such a listener interface wouldn't implement `equals` anyway) then using `Set` would be perfectly fine using the default implementation of equals/hashcode from `Object`. But regardless, any class that overrides it's `equals` method should also override its `hashcode` method as a general rule so usage of Set vs List shouldn't be concerned about how the objects it contains implement equals/hashcode as long as they follow the proper contracts.
Anyway, I only brought up using LinkedHashSet it maintains both the uniqueness contract of a Set and maintains insertion order like a List. Since you clearly care about the uniqueness of listeners but don't use index based access ability of the List, recommending a LinkedHashSet would appear to be a safe bet. If you don't even care about the insertion ordering then a plain HashSet would be just fine too.
So I guess it comes down to the question of "Are you expecting SonyModelListener implementations to not adhere to the equals/hashcode contract? If so my change request is moot.